ggml/src/ggml-vulkan/ggml-vulkan.cpp

```diff
@@ -7236,6 +7357,12 @@ static void ggml_backend_vk_buffer_free_buffer(ggml_backend_buffer_t buffer) {
     ggml_backend_vk_buffer_context * ctx = (ggml_backend_vk_buffer_context *)buffer->context;
     ggml_vk_destroy_buffer(ctx->dev_buffer);
     delete ctx;
+
+    vk_instance_buffers--;
+
+    if (vk_instance_contexts == 0 && vk_instance_buffers == 0) {
```

Comment:
If these counters really need to be atomic, then this check should use the return value from the decrement rather than separately reading the value, to avoid the situation where two threads decrement and then both read zero. But with two counters, it gets more complex and maybe a mutex would be better?

Reply:
They don't. In the beginning I had only one counter, so it made more sense. I replaced them with a mutex now.

There's still the open issue about making the Vulkan backend multithread-capable, which needs some refactoring. I'm trying not to make the situation worse when I add new features.